crates/hcl-edit/src/parser/expr.rs

```diff
@@ -390,14 +413,21 @@ fn for_object_expr<'i, 's>(
                 cut_tag("=>"),
                 decorated(ws, expr, ws),
             ),
-            opt("..."),
-            opt(for_cond),
+            opt(("...", raw_string(ws))),
+            opt(for_cond(state)),
         )
             .map(|(intro, (key_expr, value_expr), grouping, cond)| {
                 let mut expr = ForExpr::new(intro, value_expr);
                 expr.key_expr = Some(key_expr);
-                expr.grouping = grouping.is_some();
                 expr.cond = cond;
+                if let Some((_, trailing)) = grouping {
+                    expr.grouping = true;
+                    if let Some(ref mut cond) = expr.cond {
+                        cond.decor_mut().set_prefix(trailing);
+                    } else {
+                        expr.decor_mut().set_suffix(trailing);
+                    }
+                }
```

Comment:
lmk if there's a better pattern here

Reply:
I don't think there's a problem with this pattern, in fact I'm doing similar things in other areas of the parser already, both for simplicity and performance reasons.

It's a bit hard to express the conditional association of this piece of whitespace to either the cond or the for expr depending on the existence of the `...`, the cond, or both, using parser combinators.

You could probably express these cases using an `alt` combinator, but this will make the performance suffer because this will lead to useless expression parsing only to throw them away if the case does not match, and then to parse it again for the next case.

So what you did is probably the best performing and most simple implementation you could come up here I would say.